networking/src/p2p/peer.rs

```diff
@@ -206,6 +208,192 @@ struct Network {
+            (1, 1), // Disconnect
+            (1, 1), // Advertise
+            (10, 10), // SwapRequest
+            (10, 10), // SwapAck
+            (1, 1), // Bootstrap
+            (10, 500), // GetCurrentBranch
+            (200, 10), // CurrentBranch
+            (10, 10), // Deactivate
+            (10, 10), // GetCurrentHead
+            (500, 500), // CurrentHead
+            (5000, 5000), // GetBlockHeaders
+            (5000, 5000), // BlockHeader
+            (500, 5000), // GetOperations
+            (20000, 10000), // Operation
+            (10, 10), // GetProtocols
+            (10, 10), // Protocol
+            (5000, 5000), // GetOperationsForBlocks
+            (10000, 10000), // OperationsForBlocks
+        ];
+        for (i, s) in THROTTLING_QUOTA_STRS.iter().enumerate() {
+            let var = "THROTTLING_QUOTA_".to_owned() + &s.to_uppercase();
+            if let Ok(val) = std::env::var(var).or_else(|_| std::env::var("THROTTLING_QUOTA_MAX")) {
+                let q = val.split(",").collect::<Vec<_>>();
+                if q.len() == 2 {
+                    if let (Ok(tx), Ok(rx)) = (q[0].parse::<isize>(), q[1].parse::<isize>()) {
+                        default[i] = (tx, rx);
+                    }
+                }
+            }
+        }
+        default
+    };
+}
+
+struct ThrottleQuota {
+    quotas: [(isize, isize); THROTTLING_QUOTA_NUM],
+    quota_disabled: bool,
+    log: Logger,
+}
+
+impl ThrottleQuota {
+    fn new(log: Logger) -> Self {
+        Self {
+            quotas: THROTTLING_QUOTA_MAX.clone(),
+            quota_disabled: *THROTTLING_QUOTA_DISABLE,
+            log,
+        }
+    }
+
+    fn msg_index(msg: &PeerMessageResponse) -> usize {
+        debug_assert!(THROTTLING_QUOTA_NUM == 18);
+        match msg.message() {
+            PeerMessage::Disconnect => 0,
+            PeerMessage::Advertise(_) => 1,
+            PeerMessage::SwapRequest(_) => 2,
+            PeerMessage::SwapAck(_) => 3,
+            PeerMessage::Bootstrap => 4,
+            PeerMessage::GetCurrentBranch(_) => 5,
+            PeerMessage::CurrentBranch(_) => 6,
+            PeerMessage::Deactivate(_) => 7,
+            PeerMessage::GetCurrentHead(_) => 8,
+            PeerMessage::CurrentHead(_) => 9,
+            PeerMessage::GetBlockHeaders(_) => 10,
+            PeerMessage::BlockHeader(_) => 11,
+            PeerMessage::GetOperations(_) => 12,
+            PeerMessage::Operation(_) => 13,
+            PeerMessage::GetProtocols(_) => 14,
+            PeerMessage::Protocol(_) => 15,
+            PeerMessage::GetOperationsForBlocks(_) => 16,
+            PeerMessage::OperationsForBlocks(_) => 17,
+        }
+    }
+
+    fn index_to_str(index: usize) -> &'static str {
+        if index < THROTTLING_QUOTA_NUM {
+            THROTTLING_QUOTA_STRS[index]
+        } else {
+            "<invalid index>"
+        }
+    }
+
+    fn can_send(&mut self, msg: &PeerMessageResponse) -> Result<isize, isize> {
```

Comment:
@akoptelov 
I would simplify this a little bit:
`Result<isize, isize>`  -> `Result<(), isize>`
I dont see anywhere to handle `Ok(quota)`

also the same for `can_receive`